napari/utils/action_manager.py

```diff
@@ -0,0 +1,287 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from inspect import signature
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Set, Union
+
+from .interactions import Shortcut
+from .key_bindings import KeymapProvider
+
+if TYPE_CHECKING:
+    from qtpy.QtWidgets import QAction, QPushButton
+
+    from napari.qt import QtStateButton
+
+
+def call_with_context(function, context):
```

Comment:
This is a comment to keep things in mind.

To use the VScode example we talked about in previous meetings
<img width="698" alt="Screen Shot 2021-04-26 at 3 38 35 PM" src="https://user-images.githubusercontent.com/3627835/116147546-7ce30700-a6a5-11eb-97d4-6d82612e74a8.png">

Context here refers to `Where` I apply a given action or `Who` provides the given action?

We will also need to consider the `When`